### PR TITLE
cpu: add capability to show stack usage of ISR

### PR DESCRIPTION
**Pull request by MohmadAyman**
replacement of PR #4201 where its history got missed up, and i couldn't fix.
Tested on Native and Arduino-due boards, Hack and Ack for the rest of the boards.

If the board is not a Cortex based board, the output when running the 'ps' command should contain a line like this
`	       - | isr_stack            | -        - |  -  |  xxx (   -1) |   ` 
xxx could be any number depending on the cpu,
CPUs that doesn't contain an explicit 'isr_stack' , the xxx which is the isr_stacksize is set to Zero.

### CODE REVIEW

File: sys/ps/ps.c [hunk 1]
```diff
@@ -82,12 +82,29 @@ void ps(void)
...
             stacksz -= thread_measure_stack_free(p->stack_start);
+            int isr_usage = thread_arch_isr_stack_usage();                         /* ISR stack usage */
             overall_used += stacksz;
 #endif
 #ifdef MODULE_SCHEDSTATISTICS
             double runtime_ticks =  sched_pidlist[i].runtime_ticks / (double) xtimer_now() * 100;
             int switches = sched_pidlist[i].schedules;
 #endif
+
+            #ifdef DEVELHELP
```
Comment (miri64) [inline sys/ps/ps.c]: Here and below: weird indentation

File: cpu/msp430-common/include/cpu_conf.h [hunk 2]
```diff
@@ -25,7 +25,7 @@ extern "C" {
 #endif
 
 #define THREAD_STACKSIZE_IDLE      (96)
-#define MSP430_ISR_STACK_SIZE           (256)
+#define ISR_STACKSIZE      (256)
```
Comment (miri64) [inline cpu/msp430-common/include/cpu_conf.h]: Please align value with the value in the line above?

Also: Why has native it's own commit for that, but not the rest?

**Comment by miri64** [comment]
Can you split up the commits a little bit more thematically and give them somewhat more helpful summaries using `git rebase -i $(git merge-base master HEAD)` with the `edit` command (and `git commit --amend`)? Something along the lines of:

* `cpu: unify naming of ISR stacksize macro`
* `core: provide function to acquire ISR stack usage`
* ``shell: provide ISR stackusage information in `ps` command``

**Comment by MohmadAyman** [comment]
Well, I usually miss up the history when i use the rebase, but I'll try.

**Comment by miri64** [comment]
`git reflog` is your friend if you mess up ;-)

File: cpu/msp430-common/include/cpu_conf.h [hunk 3]
```diff
@@ -25,7 +25,7 @@ extern "C" {
 #endif
 
 #define THREAD_STACKSIZE_IDLE      (96)
-#define MSP430_ISR_STACK_SIZE           (256)
+#define ISR_STACKSIZE                    (256)
```
Comment (miri64) [inline cpu/msp430-common/include/cpu_conf.h]: `(256)` still not aligned to the `(96)`

File: cpu/native/include/cpu_conf.h [hunk 4]
```diff
@@ -36,7 +36,7 @@ extern "C" {
 #define THREAD_STACKSIZE_MINIMUM            (163840)
 /* native internal */
 #define THREAD_STACKSIZE_MINIMUM            (163840)
-#define NATIVE_ISR_STACKSIZE                (163840)
+#define ISR_STACKSIZE                (163840)
```
Comment (miri64) [inline cpu/native/include/cpu_conf.h]: Alignment

File: cpu/native/include/cpu_conf.h [hunk 5]
```diff
@@ -46,7 +46,7 @@ extern "C" {
 /* for core/include/thread.h */
 #define THREAD_STACKSIZE_MINIMUM            (8192)
 /* native internal */
-#define NATIVE_ISR_STACKSIZE                (8192)
+#define ISR_STACKSIZE                (8192)
```
Comment (miri64) [inline cpu/native/include/cpu_conf.h]: Alignment

File: sys/ps/ps.c [hunk 6]
```diff
@@ -82,12 +82,29 @@ void ps(void)
...
+            int isr_usage = thread_arch_isr_stack_usage();                         /* ISR stack usage */
             overall_used += stacksz;
 #endif
 #ifdef MODULE_SCHEDSTATISTICS
             double runtime_ticks =  sched_pidlist[i].runtime_ticks / (double) xtimer_now() * 100;
             int switches = sched_pidlist[i].schedules;
 #endif
+
+#ifdef DEVELHELP
+        if(i==1){
```
Comment (miri64) [inline sys/ps/ps.c]: (here and below) 4 leading spaces missing compared to surrounding code

**Comment by miri64** [comment]
@MohmadAyman This one time (!) please do the following:

```bash
git checkout isr_stack_usage
git fetch git://github.com/miri64/RIOT.git MohmadAyman/fixes
git reset --hard FETCH_HEAD
git push -f
```

**Comment by miri64** [comment]
@gebart It feels weird to ACK this now, can you have a look over it after @MohmadAyman pushed, too?

**Comment by miri64** [comment]
Thanks @MohmadAyman (please write a comment when you update a PR, the notification system of GitHub does not inform on new/changed commits)

**Comment by miri64** [comment]
I tested this configuration when I pushed the cleaned-up branch on `native` and `samr21-xpro`. As far as I'm concerned this PR can be merged.

File: cpu/cortexm_common/thread_arch.c [hunk 7]
```diff
@@ -251,6 +254,15 @@ void thread_arch_stack_print(void)
     printf("current stack size: %i byte\n", count);
 }
 
+/* This function returns the number of bytes used on the ISR stack */
+int thread_arch_isr_stack_usage(void)
+{
+    uint32_t  *ptr = &_sstack;
+    while (*(ptr++) == STACK_CANARY_WORD) {
+    }
```
Comment (OlegHahm) [inline cpu/cortexm_common/thread_arch.c]: wasn't the conclusion to use 'while(foo) {}`?

File: cpu/cortexm_common/thread_arch.c [hunk 8]
```diff
@@ -251,6 +254,15 @@ void thread_arch_stack_print(void)
...
+    return ISR_STACKSIZE - (ptr - &_sstack);
```
Comment (OlegHahm) [inline cpu/cortexm_common/thread_arch.c]: I would prefer parentheses

File: sys/ps/ps.c [hunk 9]
```diff
@@ -82,12 +82,24 @@ void ps(void)
...
+            int isr_usage = thread_arch_isr_stack_usage();                         /* ISR stack usage */
             overall_used += stacksz;
 #endif
 #ifdef MODULE_SCHEDSTATISTICS
             double runtime_ticks =  sched_pidlist[i].runtime_ticks / (double) xtimer_now() * 100;
             int switches = sched_pidlist[i].schedules;
 #endif
+
+#ifdef DEVELHELP
+            if (i == 1) {
```
Comment (OlegHahm) [inline sys/ps/ps.c]: Why `1`?

**Comment by miri64** [comment]
@OlegHahm given @MohmadAyman inexperience with git I would rather merge this and provide a fixup PR.

**Comment by OlegHahm** [comment]
Okay, only minor comments anyway - can be fixed later on. ACK

Can you also open an issue that we need implementations for the other architectures, too?

File: sys/ps/ps.c (hunk 9 above)
Comment (MohmadAyman) [inline sys/ps/ps.c]: to put the isr_size line on top.
```
	pid | name                 | state    Q | pri | stack ( used) | location   
	  - | isr_stack            | -        - |  -  |  8192 (   -1) |    
	  1 | idle                 | pending  Q |  15 |  8192 ( 6240) | 0x80621c0 
	  2 | main                 | running  Q |   7 | 12288 ( 9312) | 0x805f1c0 
	  3 | pktdump              | bl rx    _ |   6 | 12288 ( 9312) | 0x806da40 
	  4 | gnrc_netdev2_tap     | bl rx    _ |   4 |  8192 ( 6240) | 0x806b1e0 
	    | SUM                  |            |     | 49152 (31103)

```

**Comment by miri64** [comment]
Will do.

File: cpu/cortexm_common/thread_arch.c (hunk 7 above)
Comment (MohmadAyman) [inline cpu/cortexm_common/thread_arch.c]: I don't really remember which way did we settle on, but i guess this will do.

**Comment by miri64** [comment]
See #5494

**Comment by OlegHahm** [comment]
Apparently `ISR_STACKSIZE` is not defined for all platforms.

**Comment by miri64** [comment]
There are issues with `arm7`, `atmega` and `qemu-i386` according to Murdock. @MohmadAyman please do the same proceedure outlined in https://github.com/RIOT-OS/RIOT/pull/5491#issuecomment-223087036 again. I fixed it there + included @OlegHahm's proposed style changes.

**Comment by MohmadAyman** [comment]
there was some error in the ` qemu ` , the branch now is good to go, i squashed the salt commit into the ` ps ` commit, you might want to check it @miri64 .

**Comment by miri64** [comment]
Good enough for me =D re-ACK and go.

**Pull request merged**